inc/listeners/heartbeat.h

```diff
@@ -0,0 +1,91 @@
+#ifndef HEARTBEAT_H
+#define HEARTBEAT_H
+
+#include <iostream>
+#include <deque>
+#include <vector>
+#include <chrono>
+#include <fmt/chrono.h>
+
+#include "events.h"
+#include "instruction.h"
+
+class Heartbeat {
+public:
+  static constexpr auto cli_key = "Heartbeat";
+
+  uint64_t cycles_between_printouts = 10000000;
+  std::vector<uint64_t> num_retired_last_printout;
+  std::vector<uint64_t> cycles_last_printout;
+  std::vector<uint64_t> num_retired;
+  
+  std::vector<uint64_t> num_retired_start_phase;
+  std::vector<uint64_t> cycles_start_phase;
+  std::vector<bool> switched_phase; // true if there has been a begin_phase event since the last time RETIRE occurred
+
+  template<Event e, typename... Args>
+  void handle_event(Args&... args);
+  
+  void add_cpu(uint32_t cpu) {
+    while (cpu >= switched_phase.size()) {
+      num_retired_last_printout.push_back(0);
+      cycles_last_printout.push_back(0);
+      num_retired.push_back(0);
+      num_retired_start_phase.push_back(0);
+      cycles_start_phase.push_back(0);
+      switched_phase.push_back(false);
+    }
+  }
+};
+
+std::chrono::seconds elapsed_time();
+
+namespace heartbeat {
+
+template<Event e, typename... Args>
+inline void handle_event(Heartbeat* hb, Args&... args) {
+  //std::cout << "WARNING: generic handle event\n";
+}
+
+template<>
+inline void handle_event<Event::BEGIN_PHASE>(Heartbeat* hb, bool& is_warmup) {
+  for (size_t i = 0; i < hb->switched_phase.size(); i++) {
+    hb->switched_phase[i] = true;
+  }
+}
+
+template<>
+inline void handle_event<Event::RETIRE>(Heartbeat* hb, uint32_t& cpu, std::deque<ooo_model_instr>::const_iterator& begin, std::deque<ooo_model_instr>::const_iterator& end, uint64_t& current_cycles) {
+  hb->add_cpu(cpu);
+  hb->num_retired[cpu] += std::distance(begin, end);
+  
+  if (hb->switched_phase[cpu]) {
+    hb->switched_phase[cpu] = false;
+    hb->num_retired_start_phase[cpu] = hb->num_retired[cpu];
+    hb->cycles_start_phase[cpu] = current_cycles;
+  }
+  
+  if (hb->num_retired[cpu] > hb->num_retired_last_printout[cpu] + hb->cycles_between_printouts) {
+    
+    double heartbeat_instr = hb->num_retired[cpu] - hb->num_retired_last_printout[cpu];
+    double heartbeat_cycle = current_cycles - hb->cycles_last_printout[cpu];
+    
+    double phase_instr = hb->num_retired[cpu] - hb->num_retired_start_phase[cpu];
+    double phase_cycle = current_cycles - hb->cycles_start_phase[cpu];
+    
+    fmt::print("Heartbeat CPU {} instructions: {} cycles: {} heartbeat IPC: {:.4} cumulative IPC: {:.4} (Simulation time: {:%H hr %M min %S sec})\n", cpu,

```

Comment:
`fmt::print` has an overload that takes a `std::ostream`. If you make it a `std::ostringstream`, then you can view it in a test.